src/main/java/org/generationcp/middleware/dao/oms/CVTermDao.java

```diff
@@ -1381,4 +1418,331 @@ public CVTerm save(final String name, final String definition, final CvId cvId)
+			final List<String> variableTypeNames =
+				variableTypes.stream().map(i -> VariableType.getById(i).getName()).collect(Collectors.toList());
+			sqlQuery.setParameterList(VARIABLE_TYPE_NAMES, variableTypeNames);
+			sqlQuery.setParameter("datasetId", datasetId);
+			sqlQuery.setParameterList("variableTypes", variableTypes);
+
+			return ((BigInteger) sqlQuery.uniqueResult()).longValue();
+
+		} catch (final HibernateException e) {
+			LOG.error(e.getMessage(), e);
+			throw new MiddlewareQueryException("Error in countVariablesByStudy() query in CVTermDao: " + e.getMessage(), e);
+		}
+	}
+
+	/**
+	 * Gets the list of variables associated to a study filtered by variableTypes.
+	 *
+	 * @param pageSize
+	 * @param pageNumber
+	 * @param studyId
+	 * @param variableTypes
+	 * @param cropName
+	 * @return
+	 */
+	public List<VariableDTO> getVariablesByDatasetId(final Integer pageSize, final Integer pageNumber, final Integer datasetId,
+		final List<Integer> variableTypes, final String cropName) {
+
+		final SQLQuery sqlQuery = this.getSession().createSQLQuery(this.createVariableQuery(true));
+		this.appendGetVariablesScalar(sqlQuery);
+
+		final List<String> variableTypeNames =
+			variableTypes.stream().map(i -> VariableType.getById(i).getName()).collect(Collectors.toList());
+		sqlQuery.setParameterList(VARIABLE_TYPE_NAMES, variableTypeNames);
+		sqlQuery.setParameter("datasetId", datasetId);
+		sqlQuery.setParameterList("variableTypes", variableTypes);
+
+		return this.convertToVariableDTO(this.getVariableQueryResult(pageSize, pageNumber, sqlQuery), cropName, true);
+
+	}
+
+	public List<Map<String, Object>> getVariableQueryResult(final Integer pageSize, final Integer pageNumber, final SQLQuery sqlQuery) {
+		if (pageNumber != null && pageSize != null) {
+			sqlQuery.setFirstResult(pageSize * (pageNumber - 1));
+			sqlQuery.setMaxResults(pageSize);
+		}
+		sqlQuery.setResultTransformer(AliasToEntityMapResultTransformer.INSTANCE);
+		return sqlQuery.list();
+	}
+
+	private String createCountVariableQuery(final boolean isFilterByStudyId) {
+
+		final StringBuilder stringBuilder = new StringBuilder(" SELECT COUNT(DISTINCT variable.cvterm_id) ");
+		stringBuilder.append(" FROM cvterm variable  ");
+		stringBuilder.append("   INNER JOIN cvtermprop variableType ");
+		stringBuilder.append("   ON variableType.cvterm_id = variable.cvterm_id AND variableType.type_id = " + TermId.VARIABLE_TYPE
+			.getId());
+		stringBuilder.append("	  LEFT JOIN projectprop pp ON pp.variable_id = variable.cvterm_id");
+		stringBuilder.append("	  LEFT JOIN project dataset ON dataset.project_id = pp.project_id");
+		stringBuilder.append(" WHERE variableType.value in (:variableTypeNames) ");
+
+		if (isFilterByStudyId) {
+			stringBuilder.append(" AND dataset.project_id = :datasetId ");
+			stringBuilder.append(" AND pp.type_id in(:variableTypes)");
+		}
+
+		return stringBuilder.toString();
+
+	}
+
+	private String createVariableQuery(final boolean isFilterByStudyId) {
```

Comment:
Same here with `createVariableQuery `on refactoring comments for `createCountVariableQuery `